bcda/auth/oktabackend.go

```diff
@@ -0,0 +1,305 @@
+	}
+
+	return clientID, clientSecret, nil
+}
+
+type Policy struct {
+	ID          string `json:"id"`
+	Type        string `json:"type"`
+	Status      string `json:"status"`
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	Priority    int    `json:"priority"`
+	System      bool   `json:"system"`
+	Conditions  Cond   `json:"conditions"`
+}
+
+type Cond struct {
+	Clients Cli `json:"clients"`
+}
+
+type Cli struct {
+	Include []string `json:"include"`
+}
+
+// Update the Auth Server's access policy to include our new client application. Otherwise, that application
+// will not be able to use the server. To do this, we first get the current list of clients, add the new
+// server to the inclusion list, and put it back to the server
+func addClientToPolicy(clientID string) error {
+	// get the current list
+	policyUrl := fmt.Sprintf("%s/api/v1/authorizationServers/%s/policies", apiBase, authServerID)
+	req, err := http.NewRequest("GET", policyUrl, nil)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Authorization", authString)
+
+	var client = &http.Client{Timeout: time.Second * 10,}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	var result []Policy
+	err = json.NewDecoder(resp.Body).Decode(&result)
+	if err != nil {
+		return err
+	}
+
+	if len(result) > 1 {
+		return fmt.Errorf("more than one policy entry for server; can't continue safely")
+	}
+
+	// add the new client to the inclusion list
+	incl := result[0].Conditions.Clients.Include
+	incl = append(incl, clientID)
+	result[0].Conditions.Clients.Include = incl
+
+	// put the list back to the server
+	body, err := json.Marshal(result[0])
+	if err != nil {
+		return err
+	}
+
+	req, err = http.NewRequest("PUT", fmt.Sprintf("%s/%s", policyUrl, result[0].ID), bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Authorization", authString)
+
+	client = &http.Client{Timeout: time.Second * 10,}
+	resp, err = client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != 200 {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Printf("failed to update policy with %s because %v", clientID, err)
+			return err
+		}
+		return fmt.Errorf("unexpected result: %s", body)
+	}
+
+	return nil
+}
+
+func refreshKeys() {
+	for range time.Tick(time.Hour * 1) {
+		log.Println("refreshing okta public keys")
+		getPublicKeys()
+	}
+}
+
+// Gets the current set of public server signing keys. This code treats all failures as fatal,
+// since failures are likely to result in an unrecoverable state for the application.
+func getPublicKeys() map[string]rsa.PublicKey {
+	var (
+		body []byte
+		err  error
+		keys map[string]rsa.PublicKey
+	)
+
+	body, err = get(os.Getenv("OKTA_OAUTH_SERVER_METADATA"))
```

Comment:
Since we have multiple URLs being created, can we generate this one based on the base URL and admin ID env vars?